### DISCUSSION TITLE
voicebox?

### DISCUSSION THREAD

**Discussion by lucidrains** [lucidrains/soundstorm-pytorch#13] [General]
can someone knowledgeable comment on this new paper using flow matching? https://ai.facebook.com/blog/voicebox-generative-ai-model-speech/

Is it better than soundstorm?

**Comment by seastar105** [comment]
it seems Voicebox is much similar to NaturalSpeech2. both of two are fully non-autoregressive by using duration predictor, while soundstorm needs autoregressive Text-to-Semantic(T2S) transformer for TTS. so, Voicebox has more controllability than soundstorm.

Although NaturalSpeech2 predicts soundstream's latent and Voicebox predicts mel spectrogram, Voicebox also tried Encodec's pre-quantized latent and post-quantized latent as prediction target. so, that's why i said both of two are similar.

skipping and repeating was classic robustness problem in autoregressive tts model since tacotron due to its unstable attention. T2S model may not suffer from such problem, but i can say Voicebox would be better than soundstorm, at least, in terms of controllability.

**Reply by lucidrains** [reply]
thank you! i think i'll read the first flow matching paper then this weekend

**Reply by lucidrains** [reply]
@seastar105 is your assessment that NS2 would also have better controllability?

**Reply by seastar105** [reply]
@lucidrains Yes, controllability i said comes from external duration, pitch conditioning. so ns2 would also have better controllability.

for example, we can insert some break between words explicitly or speed control of utterance. since ns2 has pitch predictor, it can also inference with explicit pitch condition. this is how they synthesize singing voice.

```
To synthesize a singing voice, we use the ground-truth pitch and duration 
from another singing voice, and use different singing prompts to generate 
singing voices with different singer timbres
```

soundstorm used simple conditioning for audio generation. just prepending semantic tokens, which are derived from pretrained speech ssl model and AR T2S transformer. it can do zero-shot tts, voice conversion well. 

NS2 and Voicebox are fully NAR model. Authors show models have nice in-context-learning capability so that models can do various tasks. Voice conversion, Zero-shot TTS, denoising, speech inpainting for voicebox. 

main difference between of two are training objective and way to train duration predictor.

NS2 takes segment from utterance and predict other parts of utterance. instead, Voicebox mask segment then infilling that segment.

NS2 jointly trains audio module and duration/pitch modules, but Voicebox train audio generation module and duration module separately. maybe we can use solely trained duration module then use with soundstorm's S2A.

**Reply by lucidrains** [reply]
why can't we bring the duration and pitch conditioning into the soundstorm framework?

**Reply by seastar105** [reply]
since main contribution of soundstorm is masked generation of audio codec code, we can use it!

as you know, duration modeling is necessary due to sequence mismatching of audio sequence and text sequence. so someone can use soundstorm's S2A model with training duration module for generate semantic sequence length from text. or just conditioning masked generation with cross attention from rich text encoder(maybe T5 or PL-BERT?)

i believe it would work. since `text encoding -> predict duration -> expand to acoustic sequence dim -> decoding` this pipeline is quite general in tts.

**Reply by lucidrains** [reply]
thank you for sharing your expertise! i will continue to build out both approaches then, and probably let this repository accommodate further conditioning from the duration and pitch prediction modules from [NS2](https://github.com/lucidrains/naturalspeech2-pytorch)

**Reply by lucidrains** [reply]
will probably hold off on voicebox unless if i get excited after reading the [flow matching paper](https://arxiv.org/abs/2210.02747)

**Reply by lucidrains** [reply]
actually got a couple TTS companies reach out asking me to implement voicebox. I think soundstorm is better though; am I missing something?

**Reply by lucidrains** [reply]
i've seen so much research fall by the way side in favor of [the bitter lesson](http://www.incompleteideas.net/IncIdeas/BitterLesson.html)

unless there is something remarkable about flow matching, should just scale up. look at audiopalm..

**Reply by lucidrains** [reply]
i guess i should read the flow matching paper tomorrow, just to see if i'm missing something, or perhaps the technique is orthogonal to NAR, and can be incorporated into soundstorm

**Reply by lucidrains** [reply]
ahh ok, yeah it isn't an orthogonal concept. they claim it to be more efficient than diffusion (18 vs 250 steps) under construction now https://github.com/lucidrains/voicebox-pytorch

**Reply by Jiang-Stan** [reply]
Hi, lucidrains,
It seems that the inference speed of VoiceBox is much faster than other diffusion based methods?

Which is the method Voicebox compared to? NaturalSpeech2?

Is there any obvious strengthen or weakness between NaturalSpeech2 and Voicebox? Could you give me some insight？

Thanks a lot in advance! Hope to hear from you.
